sources/api/bootstrap-containers/src/main.rs

```diff
@@ -0,0 +1,708 @@
+const HOST_CTR_BIN: &str = "/bin/host-ctr";
+
+/// Stores user-supplied global arguments
+#[derive(Debug)]
+struct Args {
+    log_level: LevelFilter,
+    socket_path: String,
+}
+
+impl Default for Args {
+    fn default() -> Self {
+        Self {
+            log_level: LevelFilter::Info,
+            socket_path: DEFAULT_API_SOCKET.to_string(),
+        }
+    }
+}
+
+/// Stores the subcommand to be executed
+#[derive(Debug)]
+enum Subcommand {
+    CreateContainers,
+    MarkBootstrap(MarkBootstrapArgs),
+}
+
+#[derive(Debug)]
+struct MarkBootstrapArgs {
+    container_id: String,
+    mode: ValidBootstrapContainerMode,
+}
+
+/// Print a usage message in the event a bad arg is passed
+fn usage() {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!(
+        r"Usage: {} SUBCOMMAND [ ARGUMENTS... ]
+
+    Subcommands:
+        create-containers
+        mark-bootstrap
+
+    Global arguments:
+        [ --socket-path PATH ]
+        [ --log-level trace|debug|info|warn|error ]
+
+    Mark bootstrap arguments:
+        --container-id CONTAINER-ID
+        --mode MODE
+
+    Socket path defaults to {}",
+        program_name, DEFAULT_API_SOCKET,
+    );
+}
+
+/// Parses user arguments into an Args struct
+fn parse_args(args: env::Args) -> Result<(Args, Subcommand)> {
+    let mut global_args = Args::default();
+    let mut subcommand = None;
+    let mut subcommand_args = Vec::new();
+
+    let mut iter = args.skip(1);
+    while let Some(arg) = iter.next() {
+        match arg.as_ref() {
+            // Global args
+            "--log-level" => {
+                let log_level = iter.next().context(error::Usage {
+                    message: "Did not give argument to --log-level",
+                })?;
+                global_args.log_level =
+                    LevelFilter::from_str(&log_level).context(error::LogLevel { log_level })?;
+            }
+
+            "-s" | "--socket-path" => {
+                global_args.socket_path = iter.next().context(error::Usage {
+                    message: "Did not give argument to --socket-path",
+                })?
+            }
+
+            // Subcommands
+            "create-containers" | "mark-bootstrap"
+                if subcommand.is_none() && !arg.starts_with('-') =>
+            {
+                subcommand = Some(arg)
+            }
+
+            // Other arguments are passed to the subcommand parser
+            _ => subcommand_args.push(arg),
+        }
+    }
+
+    match subcommand.as_deref() {
+        // Default subcommand is 'create-containers'
+        None | Some("create-containers") => Ok((global_args, Subcommand::CreateContainers {})),
+        Some("mark-bootstrap") => Ok((global_args, parse_mark_bootstrap_args(subcommand_args)?)),
+        Some(x) => {
+            return error::Usage {
+                message: format!("Unknown subcommand '{}'", x),
+            }
+            .fail()
+        }
+    }
+}
```

Comment:
I'd prefer to just error out on an unknown subcommand.

Reply:
Done